### PR TITLE
fix(autodiff): zero-safe cumprod gradient (no division by input)

### LINKED ISSUE #3864: `cumprod` gradient produces NaN when input contains zeros
## Problem Description

The current `cumprod` autodiff implementation (see https://github.com/tracel-ai/burn/pull/3819) produces NaN gradients when the input tensor contains zeros. This is because the gradient computation uses division by the input values:

```rust
// Current implementation in burn-autodiff/src/ops/tensor.rs
let grad_times_output = B::float_mul(grad, output.clone());
let grad_result = /* reverse cumsum of grad_times_output */;
B::float_div(grad_result, input)  //  Division by zero causes NaN
```

### Example

```rust
use burn::tensor::Tensor;
use burn_autodiff::Autodiff;

let tensor = Tensor::<Autodiff<Backend>, 1>::from_data([2.0, 0.0, 3.0, 4.0], &device)
    .require_grad();
let output = tensor.clone().cumprod(0);
let grads = output.sum().backward();
let grad = tensor.grad(&grads).unwrap();

// Expected: [1.0, 32.0, 0.0, 0.0]
// Actual:   [1.0, NaN, NaN, NaN]
```

## Expected Behavior

The gradient should be computed correctly even when the input contains zeros, matching PyTorch's behavior:

```python
import torch

x = torch.tensor([2.0, 0.0, 3.0, 4.0], requires_grad=True)
y = torch.cumprod(x, dim=0)  # [2, 0, 0, 0]
y.sum().backward()
print(x.grad)  # tensor([1., 32., 0., 0.])
```

## Root Cause

The gradient formula `grad_input[i] = sum_{j>=i}(grad_output[j] * output[j] / input[i])` involves division by `input[i]`, which fails when `input[i] = 0`.

## Mathematical Solution

The correct zero-safe gradient formula is:

```
grad_input[i] = left[i] * tail[i]

where:
  left[i] = prod(input[0:i])  // exclusive left cumulative product
  tail[i] = sum_{j>=i}(grad[j] * prod(input[i+1:j+1]))
```

The key insight is that this formula **doesn't involve `input[i]` in any division**, making it well-defined even when `input[i] = 0`.

The `tail[i]` term can be computed efficiently via **reverse accumulation**:

```
tail[n-1] = grad[n-1]
tail[i] = grad[i] + input[i+1] * tail[i+1]  // for i < n-1
```

### Verification

This formula has been verified to match PyTorch's output for all test cases including:
- Zeros at the beginning, middle, and end
- Multiple zeros
- Mixed zero and non-zero values

See the ignored tests in `burn-autodiff/src/tests/cumprod.rs` for comprehensive test cases with expected values.

## Implementation Challenges

The challenge is implementing the reverse accumulation in Burn using only tensor operations. The `tail` computation has sequential dependencies where each element depends on the next:

```rust
// Pseudocode - not directly expressible in current Burn
tail[n-1] = grad[n-1]
for i in (0..n-1).rev() {
    tail[i] = grad[i] + input[i+1] * tail[i+1]
}
```

This requires one of:

1. **A scan/fold operation** - Burn currently lacks a generic scan primitive like JAX's `associative_scan` or Haskell's `scanr`
2. **Custom backend kernels** - Similar to PyTorch's approach with specialized CUDA kernels
3. **Matrix formulation** - Express the computation as matrix operations (possible but complex)

### Option 1: Add a `scan` operation to Burn (Recommended)

Add a generic scan operation to Burn's tensor API that enables reverse accumulation:

```rust
// Hypothetical API
pub trait Backend {
    fn float_scan_right<F>(
        tensor: FloatTensor<Self>,
        init: FloatTensor<Self>,
        dim: usize,
        f: F,
    ) -> FloatTensor<Self>
    where
        F: Fn(FloatTensor<Self>, FloatTensor<Self>) -> FloatTensor<Self>;
}
```

This would enable implementing not just cumprod gradients correctly, but also other advanced operations like RNNs, parallel prefix sums, etc.

### Option 2: Custom backend implementations

Implement `cumprod_backward` as a custom operation in each backend:
- **CUDA/ROCm**: Use parallel scan algorithms
- **CPU**: Use sequential loops with proper vectorization
- **WGPU**: Use compute shaders with workgroup-level scans

### Option 3: Matrix-based implementation

Reformulate the computation using matrix operations that Burn already supports. This may be less efficient but would work with existing primitives.

## References

This is a well-known challenging problem across deep learning frameworks:

### PyTorch
- Implementation: `torch/csrc/autograd/FunctionsManual.cpp` - `cumprod_backward`
- Uses both O(n) and O(n²) algorithms depending on whether second derivatives are needed
- Documentation: https://pytorch.org/cppdocs/api/function_namespaceat_1aff8f64104b6c3b3f9ac5bba2abdc26d1.html

### JAX
- **PR #2596**: "Reimplement np.cumsum and np.cumprod in terms of a parallel prefix scan"
- Changed from O(n²) to O(n log n) using `associative_scan`
- Made the operation arbitrarily differentiable
- Link: https://github.com/jax-ml/jax/pull/2596

### TensorFlow/Theano
- **TensorFlow Issue #3862**: "tf.cumprod's gradient produces nans given zeros"
- **Theano Issue #5197**: "cumprod gradient handles zero input incorrectly"
- Multiple frameworks struggled with this same problem

### Stack Overflow
- "How to compute gradient of cumprod safely": https://stackoverflow.com/questions/40916955/how-to-compute-gradient-of-cumprod-safely

## Current Workaround

The current implementation works correctly for non-zero inputs. For inputs containing zeros:
- The gradient computation produces NaN
- Tests are marked as `#[ignore]` with clear documentation
- See `burn-autodiff/src/tests/cumprod.rs` for test cases that specify the correct behavior

## Test Cases

The following test cases have been added (currently ignored) in `burn-autodiff/src/tests/cumprod.rs`:

1. `should_diff_cumprod_zero_in_middle` - Zero in the middle of the tensor
2. `should_diff_cumprod_zero_at_start` - Zero at the beginning
3. `should_diff_cumprod_zero_at_end` - Zero at the end
4. `should_diff_cumprod_multiple_zeros` - Multiple zeros

Each test includes PyTorch-verified expected gradients. Once implemented, simply remove the `#[ignore]` attribute to enable the tests.

## Related Code

- **Implementation**: `burn-autodiff/src/ops/tensor.rs:1673-1697` (CumProd backward)
- **Tests**: `burn-autodiff/src/tests/cumprod.rs:59-135` (ignored zero tests)
- **TODO comment**: `burn-autodiff/src/ops/tensor.rs:1679-1682`

### PR DESCRIPTION
**Pull request by AnayGarodia**
Fixes #3864.

## Problem

The `cumprod` backward pass computed `reverse_cumsum(grad * output) / input`, dividing by the input and producing **NaN** gradients whenever the input contains zeros. Four autodiff tests for this case were `#[ignore]`d with a TODO pointing at this issue.

## Fix

Replace the division with the standard zero-safe formula `grad_input[i] = left[i] * tail[i]`, where:

- `left[i] = prod(input[0..i])` — exclusive prefix product (built by prepending `1` along `dim`, dropping the last element, then an inclusive cumprod)
- `tail[i] = grad[i] + input[i+1] * tail[i+1]` — reverse accumulation along `dim`

This never divides by `input[i]`, so it stays finite at zeros and matches `torch.cumprod`'s backward.

## Testing

Un-ignored the four zero-input tests (`zero_in_middle`, `zero_at_start`, `zero_at_end`, `multiple_zeros`), whose expected gradients are taken from PyTorch — they now pass.

```
cargo test -p burn-backend-tests --features ndarray,std --test autodiff cumprod   # 12 passed, 0 ignored
cargo clippy -p burn-autodiff --all-targets -- -D warnings                        # clean
cargo fmt --check                                                                 # clean
```

Verified numerically against PyTorch references, e.g. input `[2,0,3,4]` → grad `[1,32,0,0]` (was NaN). The fix uses only backend-agnostic tensor ops; I ran it against the ndarray backend.

### CODE REVIEW

**Comment by Charles23R** [comment]
How does the performance compare with the previous implementation on GPU? This is probably way slower. We should at the very least keep the original, faster implementation if there are no 0s in the input, and keep a note to reimplement with a scan/fold operation eventually like mentioned in #5141

**Comment by AnayGarodia** [comment]
Good call on the perf question — the sequential recurrence was indeed the weak point, and benchmarking it on wgpu/Metal surfaced a second problem: a few hundred chained slice/mul/add ops made the fusion pass fall over (repeated `elemwise_fuse` shader-generation failures, then a stack overflow at n=1024).

So I dropped the recurrence entirely and reworked the gradient into the masked parallel form PyTorch's `cumprod_backward` uses. Each lane is split at its first zero along `dim`:

- **before** the first zero: the classic `reverse_cumsum(grad * output) / input` is exact (every divisor is nonzero),
- **at** the first zero: `reverse_cumsum(grad * cumprod(input with that zero set to 1))` evaluated at that position,
- **after** it: exactly zero.

The regions are computed with `cumsum`/`cumprod` plus elementwise masks, so it's a constant number of kernel launches for every input — no data-dependent branching, no readback, no O(n) loop, and rows with no zeros reduce to the classic formula. Multiple zeros per lane are handled (everything past the first zero is zero).

Perf on an Apple M-series GPU (Metal via wgpu, fwd+bwd+sync, 50 iters — noisy laptop numbers, treat as ballpark):

| shape | old (main) | this branch |
|---|---|---|
| 32x128, no zeros | 5.5 ms | 2.9 ms |
| 32x1024, no zeros | 3.5 ms | 9.7 ms |
| 8x4096, no zeros | 6.1 ms | 9.2 ms |
| 32x1024, with zeros | 6.6 ms (NaN grads) | 3.4 ms (correct) |

Within noise on this machine, both are flat in n; the new form adds ~8 small elementwise kernels of constant overhead. A scan/fold primitive would still be nice for other ops, but this no longer depends on it. All 12 cumprod autodiff tests pass (zero-at-start/middle/end, multiple zeros, checkpointing).

File: crates/burn-autodiff/src/ops/tensor.rs [hunk 1]
```diff
@@ -2160,33 +2160,78 @@ impl<B: Backend, C: CheckpointStrategy> FloatTensorOps<Self> for Autodiff<B, C>
...
+                    // replaces exactly that zero with one before the cumprod.
+                    let input_one = B::float_add(input, at_first.clone());
+                    let omitted =
+                        reverse_cumsum(B::float_mul(grad, B::float_cumprod(input_one, dim)));
 
-                    B::float_div(grad_result, input)
+                    B::float_add(
+                        B::float_mul(classic, before_first),
+                        B::float_mul(omitted, at_first),
+                    )
```
Comment (laggui) [inline crates/burn-autodiff/src/ops/tensor.rs]: I think we can get away with just one cumprod call by deriving output from output_one instead of computing it separately:

```rust
// Replace ONLY the first zero with 1
let input_one = B::float_add(input.clone(), at_first.clone());
// Compute cumprod only once
let output_one = B::float_cumprod(input_one, dim);
// Derive standard output: output = output_one * before_first
let output = B::float_mul(output_one.clone(), before_first.clone());

// Then, just like your approach:
let input_safe = B::float_add(input, zero_mask);
let classic = B::float_div(
    reverse_cumsum(B::float_mul(grad.clone(), output)),
    input_safe,
);
let omitted = reverse_cumsum(B::float_mul(grad, output_one));

B::float_add(
    B::float_mul(classic, before_first),
    B::float_mul(omitted, at_first),
)
```

This should be faster in your benchmarks than the current approach.

**Review by laggui — COMMENTED** [review]
Thanks for addressing the initial comments. I think we can still improve the current approach a bit.

File: crates/burn-autodiff/src/ops/tensor.rs (hunk 1 above)
Comment (AnayGarodia) [reply crates/burn-autodiff/src/ops/tensor.rs]: Applied in 97eb3002. `output_one` now feeds both the masked classic output and the omitted-product path, so the backward pass uses one cumprod instead of two. `cargo test -p burn-backend-tests --test autodiff cumprod` passes all 12 base/checkpointing cases, including the zero-position and multiple-zero regressions.

**Review by laggui — APPROVED** [review]
